src/librustc/dep_graph/graph.rs

```diff
@@ -1067,20 +1093,56 @@ impl CurrentDepGraph {
         edges: SmallVec<[DepNodeIndex; 8]>,
         fingerprint: Fingerprint
     ) -> DepNodeIndex {
-        match self.node_to_node_index.get_shard_by_value(&dep_node).lock().entry(dep_node) {
+        let (index, inserted) = match self.node_to_node_index
+                                          .get_shard_by_value(&dep_node)
+                                          .lock()
+                                          .entry(dep_node) {
             Entry::Occupied(entry) => *entry.get(),
             Entry::Vacant(entry) => {
-                let mut data = self.data.lock();
-                let dep_node_index = DepNodeIndex::new(data.len());
-                data.push(DepNodeData {
-                    node: dep_node,
-                    edges,
-                    fingerprint
-                });
+                let index = self.node_count.fetch_add(1, SeqCst);
+                // Cast to u32 to ensure we didn't overflow.
+                let index = u32::try_from(index).unwrap();
+
+                let dep_node_index = DepNodeIndex::new(index as usize);
                 entry.insert(dep_node_index);
                 dep_node_index
             }
+        };
+
+        if inserted {
+            let dep_node_data = DepNodeData {
+                node: dep_node,
+                edges,
+                fingerprint
+            };
+            let inner_index = index.as_usize() / sharded::SHARDS;
+            let mut data = self.data.get_shard_by_index(index.as_usize()).lock();
+            let len = data.len();
+            if likely!(len == inner_index) {
+                data.push(dep_node_data)
+            } else {
+                let dummy_data = DepNodeData {
+                    node: DepNode::new_no_params(DepKind::Null),
+                    edges: SmallVec::default(),
+                    fingerprint: Fingerprint::ZERO,
+                };
+                if inner_index >= len {
+                    data.extend(iter::repeat(dummy_data).take(inner_index - len + 1));
+                }
+                data[inner_index] = dep_node_data;
+            }
         }
+
+        (index, inserted)
+    }
+
+    fn data(
+        &self,
+        index: DepNodeIndex,
+    ) -> MappedLockGuard<'_, DepNodeData> {
+        LockGuard::map(self.data.get_shard_by_index(index.as_usize()).lock(), |vec| {
```

Comment:
If we made `get_shared_by_index` return a tuple, this would become

```rust
let (vec, index_in_vec) = self.data.get_shard_by_index(index.as_usize());
LockGuard::map(vec.lock(), |vec| &mut vec[index_in_vec])
```

that could even be a helper method implemented on `Sharded<Vec<T>>`